third_party/amd/lib/TritonAMDGPUTransforms/InThreadTranspose.cpp

```diff
@@ -0,0 +1,620 @@
+        result.transitiveCF.append(cfSearch.value().transitiveCF);
+      }
+    } else if (auto forOp = dyn_cast<scf::ForOp>(user)) {
+      LDBG("      for op num operands: " << forOp.getNumOperands());
+      LDBG("      for op body num operands: "
+           << forOp.getBody()->getNumArguments());
+      assert(use.getOperandNumber() >= forOp.getNumControlOperands());
+      int blockArgIdx = use.getOperandNumber() - forOp.getNumControlOperands() +
+                        forOp.getNumInductionVars();
+      auto blockArg = forOp.getBody()->getArgument(blockArgIdx);
+      auto cfSearch = traverseCFForValueUses(blockArg);
+      if (failed(cfSearch))
+        return failure();
+      result.ops.append(cfSearch.value().ops);
+      result.transitiveCF.push_back(blockArg);
+      result.transitiveCF.append(cfSearch.value().transitiveCF);
+    } else {
+      result.ops.push_back(user);
+    }
+  }
+  return result;
+}
+
+/// Look for defining operation, hopping over control flow.
+///
+/// Gather all operations of type T within one def-use hop from val,
+/// control flow constructions are not considered as an operations.
+/// \returns true on success, false if analysis failed
+template <typename Op>
+FailureOr<SmallVector<Op>> findAllDefiningOps(Value val) {
+  auto candidates = traverseCFForValueDefs(val);
+  if (failed(candidates))
+    return failure();
+  SmallVector<Op> result;
+  for (auto candidateValue : candidates.value()) {
+    auto op = candidateValue.getDefiningOp();
+    if (!op)
+      continue;
+    if (auto typedOp = dyn_cast<Op>(op))
+      result.push_back(typedOp);
+  }
+  return result;
+}
+
+/// Look for all operations with one of OpTy types in def-use chains in both
+/// forward and backward directions.
+///
+/// Traversal goes through control flow operations and and stops at non OpTy
+/// operation. For example: findAllDefUseOps<local_load, mem_subview,
+/// local_store>(dot_operand)
+///
+///                                                    ----------------->
+///                                                    local_store | traversed
+/// global_load -> local_store -> mem_subview -> local_load -> dot
+///                 traversed      traversed      traversed
+///
+/// \returns true on success, false if analysis failed
+FailureOr<loadStoreLoadPatternComponents>
+findReachableSMemOps(ttg::LocalLoadOp root,
+                     SetVector<mlir::Operation *> &visited) {
+  // breadth-first search for reachable opeations
+  loadStoreLoadPatternComponents foundNetwork;
+  SmallVector<Operation *> traversalStep{root};
+  while (!traversalStep.empty()) {
+    LDBG("begin new step in smem op analysis");
+    SmallVector<Operation *> nextTraversalStep;
+    for (auto candidate : traversalStep) {
+      if (visited.contains(candidate))
+        continue;
+      visited.insert(candidate);
+      LDBG("  processing in smem op analysis: " << *candidate);
+
+      int forwardIdx = -1;
+      int backwardIdx = -1;
```

Comment:
Can you add comment to explain what's these for to help reading the code? Also this can be `bool`? Not sure we really are indexing into something and probably naming as `bool scanForward` and `bool scanBackward` would be better?